train.py

```diff
@@ -301,11 +320,248 @@ def distributed_main(i, args, start_rank=0):
+        xm.master_print(
+            'WARNING: bfloat16 is enabled. Note that fairseq meters such as '
+            'loss will accumulate the numerator, and increment the denominator.'
+            ' Due to lack of precision in higher numbers in bfloat16, these '
+            'meters will report invalid values after a while.',
+            fd=sys.stderr
+        )
+
+    xm.master_print('Args', fd=sys.stderr)
+    for key, val in args.__dict__.items():
+        xm.master_print('\t{} {}'.format(key, val), fd=sys.stderr)
+    # `xla_device` is `torch.device` and `device` is `str`
+    xla_device = xm.xla_device()
+    task, trainer, model, epoch_itr, lr, valid_subsets, device = prepare_task(
+        args, xla_device)
+
+    train_meter = StopwatchMeter()
+    train_meter.start()
+    while keep_training(lr, epoch_itr, trainer):
+        # TRAINING
+        xm.master_print('Epoch {} begin {}'.format(epoch_itr.epoch + 1, now()))
+        progress = initialize_loader_for_epoch(
+            args, epoch_itr, prefix='training on {}'.format(device),
+        )
+        para_loader = pl.ParallelLoader(progress, [device])
+        tracker = train_loop_fn(
+            device, trainer, para_loader.per_device_loader(device),
+            len(progress) - 1
+        )
+        stats = get_training_stats(trainer)
+        progress.print(stats, tag=device)
+        print(
+            'Device {} Epoch {} Tracker Rate={:.2f}, GlobalRate={:.2f}'.format(
+                device, epoch_itr.epoch, tracker.rate(), tracker.global_rate()
+            )
+        )
+        xm.master_print('Epoch {} end {}'.format(epoch_itr.epoch, now()))
+        if args.metrics_debug:
+            xm.master_print(met.metrics_report())
+
+        # VALIDATION
+        if not args.disable_validation and epoch_itr.epoch % args.validate_interval == 0:
+            valid_losses = validate_subsets(
+                args, device, trainer, task, epoch_itr, valid_subsets
+            )
+
+            # only use average first validation loss from the first device
+            # to update the learning rate
+            vloss = valid_losses[valid_subsets[0]].item()
+            xm.master_print('old learning rate: {}'.format(lr))
+            lr = trainer.lr_step(epoch_itr.epoch, vloss)
+            xm.master_print('new learning rate: {}'.format(lr))
+        else:
+            vloss = None
+
+        # save checkpoint
+        if epoch_itr.epoch % args.save_interval == 0:
+            checkpoint_utils.save_checkpoint(args, trainer, epoch_itr, vloss)
+
+        if args.metrics_debug:
+            xm.master_print(met.metrics_report())
+
+    train_meter.stop()
+    xm.master_print('| done training in {:.1f} seconds'.format(train_meter.sum))
+    assert_on_losses(args, trainer)
+
+
+def assert_on_losses(args, trainer):
+    if xu.getenv_as('XLA_USE_BF16', bool, False):
+        # XXX: loss values are meaningless in this case due to precision in bf16
```

Comment:
What's this `XXX` about?

Reply:
this is because if bf16 is enabled, loss values will accumulate, and eventually numerator will grow large, and we'll end up with the `4096 + 10 = 4096` issue in the bf16 world.

I believe you have a pr that allows mixed precision, so this issue should be fixable. Haven't gotten around to using that yet. I may ping you when I get to that, to see how I can proceed.

Reply:
No, I mean, why `XXX` marking?

Reply:
I'll delete